src/dune/action_ast.ml

```diff
@@ -58,13 +58,45 @@ struct
               and+ args = repeat String.decode in
               Run (prog, args) )
           ; ( "with-accepted-exit-codes"
-            , Dune_lang.Syntax.since Stanza.syntax (2, 0)
+            , let open Dune_lang in
+              Syntax.since Stanza.syntax (2, 0)
               >>> let+ codes = Predicate_lang.decode_one Dune_lang.Decoder.int
-                  and+ t = located t in
-                  match t with
-                  | _, ((Run _ | Bash _ | System _) as t) ->
-                    With_accepted_exit_codes (codes, t)
-                  | loc, _ ->
+                  and+ version = Syntax.get_exn Stanza.syntax
+                  and+ loc, t = located t in
+
+                  let nesting_support =
+                    Syntax.Version.Infix.(version >= (2, 2))
+                  in
+
+                  let rec is_ok = function
+                    | Run _
+                    | Bash _
+                    | System _ ->
+                      true
+                    | Chdir (_, t)
+                    | Setenv (_, _, t)
+                    | Ignore (_, t)
+                    | Redirect_in (_, _, t)
+                    | Redirect_out (_, _, t) ->
+                      if nesting_support then
+                        is_ok t
+                      else
+                        Syntax.Error.since loc Stanza.syntax (2, 2)
```

Comment:
It might be possible to refactor that so that it uses `Syntax.since` directly but I don't think the error messages would be as precise.